// colyseus/lib/transport/WebSocketTransport.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = __importDefault(require("url"));
const __1 = require("..");
const Protocol_1 = require("../Protocol");
const Utils_1 = require("../Utils");
const Transport_1 = require("./Transport");
const Debug_1 = require("./../Debug");
function noop() { }
function heartbeat() { this.pingCount = 0; }
class WebSocketTransport extends Transport_1.Transport {
    constructor(matchMaker, options = {}, engine) {
        super(matchMaker);
        this.onConnection = (client, req) => __awaiter(this, void 0, void 0, function* () {
            // prevent server crashes if a single client had unexpected error
            client.on('error', (err) => Debug_1.debugAndPrintError(err.message + '\n' + err.stack));
            client.on('pong', heartbeat);
            // compatibility with ws / uws
            const upgradeReq = req || client.upgradeReq;
            const parsedURL = url_1.default.parse(upgradeReq.url);
            const { sessionId } = Utils_1.parseQueryString(parsedURL.query);
            const processAndRoomId = parsedURL.pathname.match(/\/[a-zA-Z0-9_\-]+\/([a-zA-Z0-9_\-]+)$/);
            const roomId = processAndRoomId && processAndRoomId[1];
            const room = this.matchMaker.getRoomById(roomId);
            // set client id
            client.pingCount = 0;
            // set client options
            client.id = sessionId;
            client.sessionId = sessionId;
            try {
                yield room._onJoin(client, upgradeReq);
            }
            catch (e) {
                Debug_1.debugAndPrintError(e.stack || e);
                Protocol_1.send[__1.Protocol.JOIN_ERROR](client, (e && e.message) || '');
                client.close(__1.Protocol.WS_CLOSE_WITH_ERROR);
            }
        });
        // disable per-message deflate
        options.perMessageDeflate = false;
        this.pingTimeout = (options.pingTimeout !== undefined)
            ? options.pingTimeout
            : 1500;
        this.pingCountMax = (options.pingCountMax !== undefined)
            ? options.pingCountMax
            : 2;
        this.wss = new engine(options);
        this.wss.on('connection', this.onConnection);
        this.server = options.server;
        if (this.pingTimeout > 0 && this.pingCountMax > 0) {
            this.autoTerminateUnresponsiveClients(this.pingTimeout, this.pingCountMax);
        }
    }
    listen(port, hostname, backlog, listeningListener) {
        this.server.listen(port, hostname, backlog, listeningListener);
        return this;
    }
    shutdown() {
        clearInterval(this.pingInterval);
        this.wss.close();
        this.server.close();
    }
    autoTerminateUnresponsiveClients(pingTimeout, pingCountMax) {
        // interval to detect broken connections
        this.pingInterval = setInterval(() => {
            this.wss.clients.forEach((client) => {
                //
                // if client hasn't responded after the interval, terminate its connection.
                //
                if (client.pingCount >= pingCountMax) {
                    return client.terminate();
                }
                client.pingCount++;
                client.ping(noop);
            });
        }, pingTimeout);
    }
}
exports.WebSocketTransport = WebSocketTransport;
